lib/Dialect/HW/Transforms/FlattenModules.cpp

```diff
@@ -32,19 +35,48 @@ using namespace igraph;
 using mlir::InlinerConfig;
 using mlir::InlinerInterface;
 
+using HierPathTable = DenseMap<hw::InnerRefAttr, SmallVector<hw::HierPathOp>>;
+
 namespace {
+
+// Cache the inner symbol attribute name to avoid repeated lookups
+static const StringRef innerSymAttrName =
+    InnerSymbolTable::getInnerSymbolAttrName();
 struct FlattenModulesPass
     : public circt::hw::impl::FlattenModulesBase<FlattenModulesPass> {
+  using FlattenModulesBase::FlattenModulesBase;
+
   void runOnOperation() override;
+
+private:
+  /// Determine if a module should be inlined based on various heuristics.
+  bool shouldInline(HWModuleOp module, igraph::InstanceGraphNode *instanceNode,
+                    size_t bodySize);
 };
 
 /// A simple implementation of the `InlinerInterface` that marks all inlining as
 /// legal since we know that we only ever attempt to inline `HWModuleOp` bodies
 /// at `InstanceOp` sites.
 struct PrefixingInliner : public InlinerInterface {
   StringRef prefix;
-  PrefixingInliner(MLIRContext *context, StringRef prefix)
-      : InlinerInterface(context), prefix(prefix) {}
+  InnerSymbolNamespace *ns;
+  HWModuleOp parentModule;
+  HWModuleOp sourceModule;
+  DenseMap<StringAttr, StringAttr> *symMapping;
+  mlir::AttrTypeReplacer *replacer;
+  HierPathTable *pathsTable;
+  hw::InnerRefAttr instanceRef;
+
+  PrefixingInliner(MLIRContext *context, StringRef prefix,
+                   InnerSymbolNamespace *ns, HWModuleOp parentModule,
+                   HWModuleOp sourceModule,
+                   DenseMap<StringAttr, StringAttr> *symMapping,
+                   mlir::AttrTypeReplacer *replacer, HierPathTable *pathsTable,
+                   hw::InnerRefAttr instanceRef)
+      : InlinerInterface(context), prefix(prefix), ns(ns),
+        parentModule(parentModule), sourceModule(sourceModule),
+        symMapping(symMapping), replacer(replacer), pathsTable(pathsTable),
+        instanceRef(instanceRef) {}
```

Comment:
This constructor may be unnecessary if you use struct initialization. E.g., `PrefixingInliner{a, b, c}`. I don't know if there's any utility in adding a one-to-one constructor here as opposed to the `{}` one.